File: math/reaching_points.kt
fun main() {
    println(reachingPoints_bfs(1,1,5,8))
    println(reachingPoints_dfs(1,1,5,8))
    println(reachingPoints_recursive(1,1,5,8))
    println(reachingPoints_backwards(1,1,5,8))
}


fun reachingPoints_bfs(sx: Int, sy: Int, tx: Int, ty: Int): Boolean {
    val queue = ArrayDeque<Pair<Int, Int>>()
    queue.addFirst(Pair<Int, Int>(sx, sy))

    while (queue.isNotEmpty()) {
        val curPair = queue.removeLast()
        if (curPair.first == tx && curPair.second == ty) {
            // we found it!
            return true
        }

        if (curPair.first > tx || curPair.second > ty) {
            // we overshot it!
            continue
        }

        // try the next two moves
        val nextPair1 = Pair<Int, Int>(curPair.first, curPair.second + curPair.first)
        queue.addFirst(nextPair1)
        val nextPair2 = Pair<Int, Int>(curPair.first + curPair.second, curPair.second)
        queue.addFirst(nextPair2)
    }

    // no path was found
    return false
}

fun reachingPoints_dfs(sx: Int, sy: Int, tx: Int, ty: Int): Boolean {
    val stack = ArrayDeque<Pair<Int, Int>>()
    stack.addFirst(Pair<Int, Int>(sx, sy))

    while (stack.isNotEmpty()) {
        val curPair = stack.removeFirst()
        if (curPair.first == tx && curPair.second == ty) {
            // we found it!
            return true
        }

        if (curPair.first > tx || curPair.second > ty) {
            // we overshot it!
            continue
        }

        // try the next two moves
        val nextPair1 = Pair<Int, Int>(curPair.first, curPair.second + curPair.first)
        stack.addFirst(nextPair1)
        val nextPair2 = Pair<Int, Int>(curPair.first + curPair.second, curPair.second)
        stack.addFirst(nextPair2)
    }

    // no path was found
    return false
}

fun reachingPoints_recursive(sx: Int, sy: Int, tx: Int, ty: Int): Boolean {
    if (sx == tx && sy == ty) {
        // we found it!
        return true
    }

    if (sx > tx || sy > ty) {
        // we overshot it!
        return false
    }

    return reachingPoints_recursive(sx, sy+sx, tx, ty) || reachingPoints_recursive(sx+sy, sy, tx, ty)
}

fun reachingPoints_backwards(sx: Int, sy: Int, tx: Int, ty: Int): Boolean {
    var curX = tx
    var curY = ty
    while (curX >= sx && curY >= sy) {
        if (curX == sx && curY == sy) {
            // we found it!
            return true
        }

        if (curX > curY) {
            curX -= curY
        } else {
            curY -= curX
        }
    }
    return false
}

File: math/reaching_points.py
#! /usr/bin/env python3

class Solution:

    def reachingPoints_modulo(self, sx: int, sy: int, tx: int, ty: int) -> bool:
        while sx < tx and sy < ty:
            tx, ty = tx % ty, ty % tx
        return sx == tx and sy <= ty and (ty - sy) % sx == 0 or \
            sy == ty and sx <= tx and (tx - sx) % sy == 0


    def reachingPoints_backwards(self, sx: int, sy: int, tx: int, ty: int) -> bool:
        x = tx
        y = ty
        while x >= sx and y >= sy:
            if x == sx and y == sy:
                return True
            if x == y:
                return False
            if y > x:
                y -= x
            elif x > y:
                x -= y
        return False


    def reachingPoints_bfs(self, sx: int, sy: int, tx: int, ty: int) -> bool:
        if sx < 1 or sy < 1 or tx < 1 or ty < 1:
            return False

        queue = []
        queue.append((sx, sy))
        while len(queue) > 0:
            cell = queue.pop(0)
            if cell[0] == tx and cell[1] == ty:
                return True
            if cell[0]+cell[1] <= tx:
                queue.append((cell[0]+cell[1], cell[1]))
            if cell[0]+cell[1] <= ty:
                queue.append((cell[0], cell[1]+cell[0]))
        return False

    def reachingPoints_dfs(self, sx: int, sy: int, tx: int, ty: int) -> bool:
        if sx == tx and sy == ty:
            return True
        if sx > tx or sy > ty:
            return False
        return self.reachingPoints_dfs(sx+sy, sy, tx, ty) or \
            self.reachingPoints_dfs(sx, sx+sy, tx, ty)


if __name__ == "__main__":
    s = Solution()
    sx = 1
    sy = 1
    tx = 7
    ty = 9
    print(s.reachingPoints_dfs(sx,sy,tx,ty))
    print(s.reachingPoints_bfs(sx,sy,tx,ty))
    print(s.reachingPoints_backwards(sx,sy,tx,ty))
    print(s.reachingPoints_modulo(sx,sy,tx,ty))
